### PR TITLE
TAS: introduce sanity e2e test

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

To make sure TAS works e2e, including computing the assignment and ungatibg the pods.

#### Which issue(s) this PR fixes:

Part of #2724 

#### Special notes for your reviewer:

I'm happy to add more scenarios in this PR or on follow ups.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by mimowo** [comment]
/test pull-kueue-test-multikueue-e2e-main
again: go: sigs.k8s.io/kind@v0.24.0: sigs.k8s.io/kind@v0.24.0: Get "https://proxy.golang.org/sigs.k8s.io/kind/@v/v0.24.0.zip": dial tcp 142.250.191.241:443: connect: no route to host

**Comment by tenzen-y** [comment]
I will pass the first review to @gabesaba

File: test/e2e/singlecluster/tas_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+			ginkgo.By(fmt.Sprintf("await for admission of workload %q and verify TopologyAssignment", wlLookupKey), func() {
+				gomega.Eventually(func() bool {
+					if err := k8sClient.Get(ctx, wlLookupKey, createdWorkload); err != nil {
+						return false
+					}
+					return createdWorkload.Status.Admission != nil
+				}, util.LongTimeout, util.Interval).Should(gomega.BeTrue())
+				gomega.Expect(createdWorkload.Status.Admission.PodSetAssignments).Should(gomega.HaveLen(1))
+				gomega.Expect(createdWorkload.Status.Admission.PodSetAssignments[0].TopologyAssignment).ShouldNot(gomega.BeNil())
```
Comment (gabesaba) [inline test/e2e/singlecluster/tas_test.go:130]: nit: is this line necessary, since we are checking the state in the next assertion?

**Review by gabesaba — COMMENTED** [review]
/lgtm

File: test/e2e/singlecluster/tas_test.go (hunk 1 above)
Comment (mimowo) [reply test/e2e/singlecluster/tas_test.go]: nope, it was just a sanity check line I added while working on the test, dropped now

File: test/e2e/singlecluster/tas_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+				gomega.Expect(createdWorkload.Status.Admission.PodSetAssignments[0].TopologyAssignment).Should(gomega.BeComparableTo(
+					&kueue.TopologyAssignment{
+						Levels: []string{
+							topologyLevel,
+						},
+						Domains: []kueue.TopologyDomainAssignment{
+							{
+								Count: 1,
+								Values: []string{
+									"kind-worker",
```
[the remark this answers is not in the document]
Comment (mimowo) [reply test/e2e/singlecluster/tas_test.go:139]: I check experimentally that the prefix comes from the cluster name, so I created a new kind cluster `mycluster`, and the nodes are named: `mycluster-worker`, `mycluster-worker2`. So the index must also be added by kind.

**Comment by mimowo** [comment]
I consider this a sanity test to verify basic scenarios (maybe also we could extend it in the future to cover https://github.com/kubernetes-sigs/kueue/pull/3292), but ultimately I think for Beta it would be have a dedicated CI Job just for TAS where we can setup a more complex node structure. Similarly as we have a dedicated CI job for multikueue: `pull-kueue-test-multikueue-e2e-main`. WDYT @tenzen-y @PBundyra ?

**Comment by tenzen-y** [comment]
> I consider this a sanity test to verify basic scenarios. Most importantly to test that the assignment is set and the topology ungater lets the Pods to start and complete. Maybe also we could extend it in the future to cover some cases in #3292.
> 
> However, ultimately I think for Beta it would be have a dedicated CI Job just for TAS where we can setup a more complex node structure. Similarly as we have a dedicated CI job for multikueue: `pull-kueue-test-multikueue-e2e-main`. WDYT @tenzen-y @PBundyra ?

That makes sense. We will set up multiple Nodes with different node labels to simulate TAS situations, right?

**Comment by mimowo** [comment]
> That makes sense. We will set up multiple Nodes with different node labels to simulate TAS situations, right?

Yes, we can figure out details later, but I think we could have, for example, 8 nodes divided into 2 blocks and 4 racks.

**Comment by tenzen-y** [comment]
> > That makes sense. We will set up multiple Nodes with different node labels to simulate TAS situations, right?
> 
> Yes, we can figure out details later, but I think we could have, for example, 8 nodes divided into 2 blocks and 4 racks.
> 
> EDIT: added the point to the [spreadsheet](https://docs.google.com/spreadsheets/d/1MXCjKZtAfqBTb61bJo46u7jIUqRIlu1NrYj1L8Xz-UU/edit?resourcekey=0-gr1ML2A1Axi8s6Lxr-Zhlw&gid=0#gid=0). I will also update the KEP after alpha.

That sounds great.

File: test/e2e/singlecluster/tas_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	workloadjob "sigs.k8s.io/kueue/pkg/controller/jobs/job"
+	"sigs.k8s.io/kueue/pkg/util/testing"
+	testingjob "sigs.k8s.io/kueue/pkg/util/testingjobs/job"
+	"sigs.k8s.io/kueue/pkg/workload"
+	"sigs.k8s.io/kueue/test/util"
+)
+
+// +kubebuilder:docs-gen:collapse=Imports
```
Comment (tenzen-y) [inline test/e2e/singlecluster/tas_test.go:39]:
```suggestion
```

We never used kubebuilder scaffolding.

File: test/e2e/singlecluster/tas_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			})
+			wlLookupKey := types.NamespacedName{Name: workloadjob.GetWorkloadNameForJob(sampleJob.Name, sampleJob.UID), Namespace: ns.Name}
+
+			ginkgo.By(fmt.Sprintf("await for admission of workload %q and verify TopologyAssignment", wlLookupKey), func() {
+				gomega.Eventually(func() bool {
+					if err := k8sClient.Get(ctx, wlLookupKey, createdWorkload); err != nil {
+						return false
+					}
+					return createdWorkload.Status.Admission != nil
+				}, util.LongTimeout, util.Interval).Should(gomega.BeTrue())
```
Comment (tenzen-y) [inline test/e2e/singlecluster/tas_test.go:128]:
```suggestion
				gomega.Eventually(func(g gomega.Gomega) {
					g.Expect(k8sClient.Get(ctx, wlLookupKey, createdWorkload)).Should(gomega.Succeeded())
					g.Expect(createdWorkload.Status.Admission.ShouldNot(gomega.BeNil())
				}, util.LongTimeout, util.Interval).Should(gomega.Suceeded())
```

Could we improve the debuggability?

File: test/e2e/singlecluster/tas_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+								Values: []string{
+									"kind-worker",
+								},
+							},
+						},
+					},
+				))
+			})
+
+			ginkgo.By(fmt.Sprintf("verify the workload %q gets finished", wlLookupKey), func() {
```
Comment (tenzen-y) [inline test/e2e/singlecluster/tas_test.go:147]: ditto

**Review by tenzen-y — COMMENTED** [review]
Basically, lgtm

File: test/e2e/singlecluster/tas_test.go (hunk 4 above)
Comment (mimowo) [reply test/e2e/singlecluster/tas_test.go]: For sure, I just copy-pasted it from somewhere, but this pattern is much better for debugability.
We can revisit also the other old places in a follow up.

File: test/e2e/singlecluster/tas_test.go (hunk 5 above)
Comment (mimowo) [reply test/e2e/singlecluster/tas_test.go:143]: done

File: test/e2e/singlecluster/tas_test.go (hunk 3 above)
Comment (mimowo) [reply test/e2e/singlecluster/tas_test.go]: done

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/lgtm
/approve
/hold cancel

**Pull request merged**